packages/playwright-core/src/remote/playwrightConnection.ts

```diff
@@ -14,63 +14,35 @@
  * limitations under the License.
  */
 
-import { SocksProxy } from '../server/utils/socksProxy';
 import { DispatcherConnection, PlaywrightDispatcher, RootDispatcher } from '../server';
 import { AndroidDevice } from '../server/android/android';
 import { Browser } from '../server/browser';
 import { DebugControllerDispatcher } from '../server/dispatchers/debugControllerDispatcher';
-import { serverSideCallMetadata } from '../server/instrumentation';
-import { assert } from '../utils/isomorphic/assert';
-import { isUnderTest } from '../server/utils/debug';
 import { startProfiling, stopProfiling } from '../server/utils/profiler';
-import { monotonicTime } from '../utils';
+import { monotonicTime, Semaphore } from '../utils';
 import { debugLogger } from '../server/utils/debugLogger';
+import { PlaywrightDispatcherOptions } from '../server/dispatchers/playwrightDispatcher';
 
 import type { DispatcherScope, Playwright } from '../server';
-import type { LaunchOptions } from '../server/types';
 import type { WebSocket } from '../utilsBundle';
-import type * as channels from '@protocol/channels';
-
-export type ClientType = 'controller' | 'launch-browser' | 'reuse-browser' | 'pre-launched-browser-or-android';
-
-type Options = {
-  allowFSPaths: boolean,
-  socksProxyPattern: string | undefined,
-  browserName: string | null,
-  launchOptions: LaunchOptions,
-  sharedBrowser?: boolean,
-};
-
-type PreLaunched = {
-  browser?: Browser | undefined;
-  androidDevice?: AndroidDevice | undefined;
-  socksProxy?: SocksProxy | undefined;
-};
 
 export class PlaywrightConnection {
   private _ws: WebSocket;
-  private _onClose: () => void;
+  private _semaphore: Semaphore;
   private _dispatcherConnection: DispatcherConnection;
   private _cleanups: (() => Promise<void>)[] = [];
   private _id: string;
   private _disconnected = false;
-  private _playwright: Playwright;
-  private _preLaunched: PreLaunched;
-  private _options: Options;
   private _root: DispatcherScope;
   private _profileName: string;
 
-  constructor(lock: Promise<void>, clientType: ClientType, ws: WebSocket, options: Options, playwright: Playwright, preLaunched: PreLaunched, id: string, onClose: () => void) {
+  constructor(semaphore: Semaphore, ws: WebSocket, controller: boolean, playwright: Playwright, initialize: () => Promise<PlaywrightDispatcherOptions & { cleanups: (() => Promise<void>)[] }>, id: string) {
```

Comment:
Let's return a single `dispose()` instead of `cleanups` array. This should simplify things.